lib/nimble_hex/mirror/server.ex

```diff
@@ -0,0 +1,151 @@
+defmodule NimbleHex.Mirror.Server do
+  @moduledoc false
+
+  use GenServer
+  require Logger
+
+  def start_link(options) do
+    {mirror, options} = Keyword.pop(options, :mirror)
+    GenServer.start_link(__MODULE__, mirror, options)
+  end
+
+  @impl true
+  def init(mirror) do
+    {:ok, mirror, {:continue, :sync}}
+  end
+
+  @impl true
+  def handle_continue(:sync, mirror) do
+    handle_info(:sync, mirror)
+  end
+
+  @impl true
+  def handle_info(:sync, mirror) do
+    sync(mirror)
+    schedule_sync(mirror)
+    {:noreply, mirror}
+  end
+
+  defp schedule_sync(mirror) do
+    Process.send_after(self(), :sync, mirror.sync_interval)
+  end
+
+  defp sync(mirror) do
+    # TODO: currently sync is naive and downloads everything everytime.
+
+    config = %{
+      :hex_core.default_config()
+      | repo_name: mirror.mirror_name,
+        repo_url: mirror.mirror_url,
+        repo_public_key: mirror.public_key,
+        http_user_agent_fragment: user_agent_fragment()
+    }
+
+    with {:ok, names} when is_list(names) <- sync_names(mirror, config),
+         {:ok, versions} when is_list(versions) <- sync_versions(mirror, config) do
+      for %{name: name} <- names,
+          !mirror.only or name in mirror.only do
+        sync_package(mirror, config, name)
+      end
+
+      for %{name: name, versions: versions} <- versions,
+          !mirror.only or name in mirror.only,
+          version <- versions do
+        sync_release(mirror, config, name, version)
+      end
+
+      # TODO: sync docs?
```

Comment:
:eyes:

Reply:
That's a good point, agreed.

FWIW even though it's kind of implementation detail of hex.pm that there happens to be `repo.hex.pm/docs/<package>-<version>.tar.gz`, I'd consider lifting that to the spec and implement appropriate APIs in hex_core. We kinda set a precedent for adding something like this by considering https://github.com/hexpm/hex_core/issues/61.

Anyhow, I'm fine with leaving this out for now, will start a discussion about this elsewhere.